### DISCUSSION TITLE
Option<i32> as foreign key part 2

### DISCUSSION THREAD

**Discussion by ixolius** [diesel-rs/diesel#3198] [Q&A] [answered]
This question is closely related to #2852 , but far enough to have its own discussion.

Suppose the following structs:
```
struct User {
  id: Option<i32>,
  //around 10 other fields..
}
struct Ticket {
  id: Option<i32>,
//around 10 other fields...
  responsible_user: Option<i32>,
}
```
In this case (and in contrast to #2852 ), `responsible_user` can be `None`, this simply means that the ticket has not been assigned yet.
Now, if I query `Ticket::belonging_to(&users).load::<Ticket>(&conn).grouped_by(&users)`, I would know what to do with Tickets with no User: simply drop and ignore them, since we obviously want only the tickets belonging to someone.

In practice, this fails, because `&Option<i32>` does not implement `Borrow<i32>`, which is required by `grouped_by`. But it would be really nice to have such a possibility. Again, it would only work, if both the foreign and primary key are optional.

Is there already a possibility to realize this? If not, would it be complicated to add one?

If my idea is a bad idea (it may be), it would however be nice to document this restriction on associations somewhere. There are plenty of code examples out there, who use `Option<i32>` as `id`, because it is a lot simpler. So a small warning in the `associations` Module documentation would be a good idea.

**Comment by weiznich** [comment, marked as answer]
`Ticket::belonging_to(&users)` is equivalent to `tickets::responsible_user.eq_any(list_of_user_ids)`, so if your user id is an `Option` that list will just contain a `None` value there, which is then send as `NULL` to the database. The `IN` operator then ignores `NULL` values.

> If my idea is a bad idea (it may be), it would however be nice to document this restriction on associations somewhere. There are plenty of code examples out there, who use Option<i32> as id, because it is a lot simpler. 

Can you point to such an example? Because the official documentation does not use optional primary keys anywhere.

> In practice, this fails, because &Option<i32> does not implement Borrow<i32>, which is required by grouped_by. But it would be really nice to have such a possibility. Again, it would only work, if both the foreign and primary key are optional.

I think that's not a good idea because the behaviour of such an implementation would not be clear at all. Additionally I do not see any straight forward way to integrate this with the existing `GroupedBy` impl without breaking other uses.

**Reply by ixolius** [reply]
> Can you point to such an example? Because the official documentation does not use optional primary keys anywhere.

Sure: The Rocket framework uses optional primary keys in its examples, because it is much easier to handle: https://github.com/SergioBenitez/Rocket/blob/v0.5-rc/examples/todo/src/task.rs

I got around this problem by copy-pasting the `grouped_by` function into a new fuction and adapting it so it allows optional ids. But thanks for your reply and thoughts.

**Reply by weiznich** [reply]
Might be worth to raise that as issue at the rocket repository as that's their documentation, not diesels.
